Cjg/julia7

Cody-G opened the pull request:
This finishes the update to Julia 0.7 /1.0 and adding of NIDAQmx 18.6.0 support.  I've tested this on the USB 6002 hardware, but I'd like to test on one of our more complex DAQ boards before we merge.  Also I still need to update the README.

Cody-G commented:
I've tested this on a couple of DAQ boards and it seems to be working well now.  @bjarthur maybe it's best that you also run the tests on your machines before merging.

One of my commits resolves an intermittent error I was getting due to passing a `Ptr` instead of a `Ref`.  I think there are other places in the NIDAQ.jl code (not the wrapped C functions but the Julia interface functions) that still pass pointers, but I haven't yet run into any errors caused by them so I left them alone.

bjarthur commented on Project.toml:
```diff
@@ -4,6 +4,7 @@ version = "0.3.0"
 
 [deps]
 Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+LinearAlgebra = "37e2e46d-f89d-539d-b4ee-838fcccc9c8e"
```
so as not to clutter the git log, could you please rebase 161a6a84 with bb2557e?

bjarthur commented on src/NIDAQ.jl:
```diff
@@ -20,10 +20,10 @@ More examples are in tests/.
 
 # tasks
-export stop, clear
+export start, stop, clear
```
i'd suggest to continue to export `start` so than it does not have to be qualifed by `NIDAQ.`

bjarthur commented on test/runtests.jl:
```diff
@@ -38,16 +38,16 @@ else
     @test typeof(getproperties(t,dev*"/ai0")) == Dict{String,Tuple{Any,Bool}}
     @test typeof(t) == NIDAQ.AITask
-    @test start(t) == nothing
+    @test NIDAQ.start(t) == nothing
```
qualifying with `NIDAQ` would not be needed if start was still exported.  does that not work for some reason?

Cody-G replied:
Makes sense, not sure why I didn't export it. I'll try it and make sure I'm not forgetting some problem with it.

Cody-G replied:
It turns out I did export `start`, see [this line](https://github.com/Cody-G/NIDAQ.jl/blob/cjg/julia7/src/NIDAQ.jl#L26).  The reason I qualified the calls to start is to avoid an ambiguity error in Julia 0.7 (`Base.start` still exists in 0.7, though it's deprecated.)

bjarthur replied:
if there is not a way to use Compat to avoid the warnings on 0.7 then i just wouldn't worry about it and leave `start` unqualified.  1.0 is not just the future, it's the present.

Cody-G replied:
I don't see a way to do this using Compat.  I can just leave it unqualified, and this will cause tests to fail on 0.7.  But the package will still work on 0.7, so I'd rather not bump the minimum version just yet (0.7 still comes in handy for us).  Sound good?

bjarthur commented on test/runtests.jl:
```diff
@@ -89,12 +89,11 @@ else
     @test stop(t) == nothing
     @test digital_input(t, dev*"/Port0/Line1") == nothing
     @test NIDAQ.start(t) == nothing
-    @test length(read(t, UInt8, 6)) == 12
```
in 4e8ca31e8f you have a nice switch to exclude tests for boards that don't have clocks.  could we do something like that here?  somehow test `read` on boards no matter how many digital lines they have??

Cody-G replied:
If I recall correctly the problem here is `UInt8`.  I think I can change all `UInt8` reads to `UInt32` reads (and update the value we're comparing to in the test) and that should be safe for any device.  Is that what you mean?  Or do you mean to skip the test if the device has more than 8 lines in a port?

bjarthur replied:
changing to `UInt32` would be great if it works.  if not, add a switch to use `UInt8` or `UInt32` as appropriate.

Cody-G replied:
UInt32 seems to work fine, though I've only been able to test it on one device.  I can't access other devices until 2 weeks from now.  If you test and it works for you feel free to merge before then (pushing the new code now).

bjarthur commented on test/runtests.jl:
```diff
@@ -117,7 +116,6 @@ else
     @test stop(t) == nothing
     @test digital_output(t, dev*"/Port0/Line1") == nothing
     @test NIDAQ.start(t) == nothing
-    @test write(t, round.(UInt8, [1 0; 0 0; 1 0; 0 1; 1 1; 0 1])) == 6
```
ditto for making this test work no matter how many digital lines the board has

bjarthur commented on src/analog.jl:
```diff
@@ -71,7 +71,7 @@ function read(t::AITask, precision::DataType, num_samples_per_chan::Integer = -1
     num_samples_per_chan_read = Int32[0]
     buffer_size = num_samples_per_chan==-1 ? 1024 : num_samples_per_chan
-    data = Array{precision}(buffer_size*num_channels)
+    data = Array{precision}(undef, buffer_size*num_channels)
```
commits 63417e, 59fc2734, 6e24dba, d05c5f13, 91c5839fd3, 9ed588b5, and d8baa7665 all appear to address deprecation warnings.  if so, so as not to clutter the git log, could you please rebase this into a single commit with the message "fix dep warns" or similar?

bjarthur commented:
looks great!  thanks for taking the time to tackle this.

Cody-G commented:
Thanks for the review!  I'm out of town at the moment but I'll make those changes asap.

Cody-G commented:
I've combined all update-related commits and I think I've addressed the comments.  It seems ready to me, but it may be best to test on one of your machines before merging.

bjarthur commented:
looks great!  thanks for the changes.  just two more suggestions:

1. about qualifying `start` with `NIDAQ`:  if that's what it takes to get 0.7 tests to pass, then let's do it.   sorry to ask you to change it back.  maybe add a note to the first use of the qualification in the tests to remind us to remove it once we drop support for 0.7, and to let users who are using the tests as documentation know that it's not necessary for 1.0.

2. it's probably safest to change all calls to `pointer` to `Ref`.  the julia docs say this is [preferable](https://docs.julialang.org/en/latest/base/c/#Base.pointer).  there might be other intermittent errors that we have just not come across yet.

i can take care of both of these if you're tired of working on this PR.

i've verified that this PR works as is on my NI-DAQ 6366.  maybe we wait to merge until after you've had a chance to test on your other devices in two weeks.

bjarthur commented:
actually, re. `start`, here's a better solution:

```
if VERSION < v"1.0.0"
  import Base.start
else
  export start
end
```

Cody-G commented:
>actually, re. `start`, here's a better solution:

Makes sense, I went that route.

> it's probably safest to change all calls to `pointer` to `Ref`.  the julia docs say this is [preferable](https://docs.julialang.org/en/latest/base/c/#Base.pointer).  there might be other intermittent errors that we have just not come across yet

Agreed.  I just didn't want to get too aggressive before I knew your opinion.  I've just pushed changes that address both points.  There were a few instances of Ptr / pointer that it didn't seem wise to change but I got most of them.  Those that I didn't change should be memory safe.

Cody-G commented:
I've tested this on our other machine and it passes.  I'd say it's ready to merge and tag a new version.

bjarthur commented:
i merged this manually on the command line so i could modify one of my commits and add another.  thanks for seeing this through!

bjarthur commented:
re. Project.toml see https://discourse.julialang.org/t/require-vs-project-toml/21962/2

Cody-G commented:
Cool thanks for merging!  And makes sense about Project.toml, now I see why most project still only have a REQUIRE.

bjarthur commented:
https://github.com/JuliaLang/METADATA.jl/pull/22229
